spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/PubsubAdmin.java

```diff
@@ -0,0 +1,151 @@
+	}
+
+	public PubsubAdmin(GcpProjectIdProvider projectIdProvider, TopicAdminClient topicAdminClient,
+			SubscriptionAdminClient subscriptionAdminClient) {
+		Assert.notNull(projectIdProvider, "The project ID provider can't be null.");
+		Assert.notNull(topicAdminClient, "The topic administration client can't be null");
+		Assert.notNull(subscriptionAdminClient,
+				"The subscription administration client can't be null");
+
+		this.projectId = projectIdProvider.getProjectId();
+		Assert.hasText(this.projectId, "The project ID can't be null or empty.");
+		this.topicAdminClient = topicAdminClient;
+		this.subscriptionAdminClient = subscriptionAdminClient;
+	}
+
+	/**
+	 * Creates a new topic on Google Cloud Pub/Sub.
+	 *
+	 * @param topicName the name for the new topic
+	 * @return the created topic
+	 */
+	public Topic createTopic(String topicName) {
+		Assert.hasText(topicName, "No topic name was specified.");
+
+		return this.topicAdminClient.createTopic(TopicName.create(this.projectId, topicName));
+	}
+
+	public void deleteTopic(String topicName) {
+		Assert.hasText(topicName, "No topic name was specified.");
+
+		this.topicAdminClient.deleteTopic(TopicName.create(this.projectId, topicName));
+	}
+
+	public Subscription createSubscription(String subscriptionName, String topicName) {
+		return createSubscription(subscriptionName, topicName, null, null);
+	}
+
+	public Subscription createSubscription(String subscriptionName, String topicName,
+			Integer ackDeadline) {
+		return createSubscription(subscriptionName, topicName, ackDeadline, null);
+	}
+
+	public Subscription createSubscription(String subscriptionName, String topicName,
+			String pushEndpoint) {
+		return createSubscription(subscriptionName, topicName, null, pushEndpoint);
+	}
+
+	/**
+	 * Creates a new subscription on Google Cloud Pub/Sub.
+	 *
+	 * @param subscriptionName the name of the new subscription
+	 * @param topicName the name of the topic being subscribed to
+	 * @param ackDeadline deadline in seconds before a message is resent. If not provided, set to
+	 *                    default of 10 seconds
+	 * @param pushEndpoint URL of the service receiving the push messages. If not provided, set to
+	 *                    Google Cloud Pub/Sub default
+	 * @return the created subscription
+	 */
+	public Subscription createSubscription(String subscriptionName, String topicName,
+			Integer ackDeadline, String pushEndpoint) {
+		Assert.hasText(subscriptionName, "No subscription name was specified.");
+		Assert.hasText(topicName, "No topic name was specified.");
+
+		int finalAckDeadline = DEFAULT_ACK_DEADLINE;
+		if (ackDeadline != null) {
+			Assert.isTrue(ackDeadline >= 0,
+					"The acknowledgement deadline value can't be negative.");
+			finalAckDeadline = ackDeadline;
+		}
+
+		PushConfig.Builder pushConfigBuilder = PushConfig.newBuilder();
+		if (pushEndpoint != null) {
+			Assert.isTrue(new UrlValidator().isValid(pushEndpoint),
```

Comment:
I share your pain with the extra dependencies for seemingly little benefit..

`new URL()` doesn't really cut it because it requires us to catch an exception instead of just returning a boolean (and the consequent performance hit, although that's not really important) and it doesn't perform validation by itself as [evidenced by this comment](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/net/URL.java#407).

Looking around the internet, this seems to be the most popular way of validating URLs.

We can also just let the Pubsub library blow up on invalid endpoint. I just tested and it does complain:

```
Exception in thread "main" com.google.api.gax.grpc.ApiException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Invalid push endpoint given (endpoint=asdasd). Refer to https://cloud.google.com/pubsub/subscriber#create for more information.
	at com.google.api.gax.grpc.ExceptionTransformingCallable$ExceptionTransformingFuture.onFailure(ExceptionTransformingCallable.java:108)
...
```